src/ReducedModelControllers/src/CentroidalMPC.cpp

```diff
@@ -0,0 +1,1341 @@
+            concatenateInput(contact.upperLimitPosition,
+                             "contact_" + key + "_upper_limit_position");
+            concatenateInput(contact.lowerLimitPosition,
+                             "contact_" + key + "_lower_limit_position");
+
+            // this only if warm-start is enabled
+            if (this->optiSettings.isWarmStartEnabled)
+            {
+                concatenateInput(contact.position, "contact_" + key + "_position_warmstart");
+            }
+
+            concatenateOutput(contact.isEnabled, "contact_" + key + "_is_enable");
+            concatenateOutput(contact.position, "contact_" + key + "_position");
+            concatenateOutput(contact.orientation, "contact_" + key + "_orientation");
+
+            std::size_t cornerIndex = 0;
+            for (const auto& corner : contact.corners)
+            {
+                concatenateOutput(corner.force,
+                                  "contact_" + key + "_corner_" + std::to_string(cornerIndex)
+                                      + "_force");
+                cornerIndex++;
+            }
+        }
+
+        casadi::Dict toFunctionOptions;
+        if (casadiVersionIsAtLeast360())
+        {
+            toFunctionOptions["cse"] = this->optiSettings.isCseEnabled;
+        }
+
+        return this->opti
+            .to_function("controller", input, output, inputName, outputName, toFunctionOptions);
+    }
+};
+
+bool CentroidalMPC::initialize(std::weak_ptr<const ParametersHandler::IParametersHandler> handler)
+{
+    constexpr auto errorPrefix = "[CentroidalMPC::initialize]";
+    auto ptr = handler.lock();
+
+    if (ptr == nullptr)
+    {
+        log()->error("{} The parameter handler is not valid.", errorPrefix);
+        return false;
+    }
+
+    if (!m_pimpl->loadParameters(ptr))
+    {
+        log()->error("{} Unable to load the parameters.", errorPrefix);
+        return false;
+    }
+
+    m_pimpl->resizeControllerInputs();
+    m_pimpl->controller = m_pimpl->createController();
+    m_pimpl->fsm = Impl::FSM::Initialized;
+
+    return true;
+}
+
+CentroidalMPC::~CentroidalMPC() = default;
+
+CentroidalMPC::CentroidalMPC()
+{
+    m_pimpl = std::make_unique<Impl>();
+}
+
+const CentroidalMPCOutput& CentroidalMPC::getOutput() const
+{
+    return m_pimpl->output;
+}
+
+bool CentroidalMPC::isOutputValid() const
+{
+    return m_pimpl->fsm == Impl::FSM::OutputValid;
+}
+
+bool CentroidalMPC::advance()
+{
+    constexpr auto errorPrefix = "[CentroidalMPC::advance]";
+    assert(m_pimpl);
+
+    using Sl = casadi::Slice;
+
+    if (m_pimpl->fsm == Impl::FSM::Idle)
+    {
+        log()->error("{} The controller is not initialized please call initialize() method.",
+                     errorPrefix);
+        return false;
+    }
+
+    // invalidate the output
+    m_pimpl->fsm = Impl::FSM::OutputInvalid;
+
+    // compute the output
+    auto controllerOutput = m_pimpl->controller(m_pimpl->vectorizedOptiInputs);
```

Comment:
In case of failure, does it throw an exception? Is it worth to catch it?

Reply:
As far as I know, there is no way to detect the failure of the optimization problem when the opti object is converted into a function.